packages/snaps-execution-environments/src/webview/WebviewSnapExecutor.ts

```diff
@@ -0,0 +1,164 @@
+ * The #jobs field is a record (an object with string keys and IJob values) that stores the jobs managed by the executor.
+ * The #proxyService field is an instance of ProxyMessageStream used for communication.
+ *
+ * The initialize static method is a factory method that creates a new instance of WebviewSnapExecutor.
+ * It takes a ProxyMessageStream as an argument and passes it to the constructor of WebviewSnapExecutor.
+ *
+ * The constructor of WebviewSnapExecutor takes an object with a proxyService property as an argument.
+ * It initializes the #jobs field as an empty object and the #proxyService field with the provided proxyService.
+ *
+ * The #onData method handles incoming messages from the WebviewExecutionService.
+ * It expects a jobId and a JSON-RPC request in the data property of the message.
+ * If the job with the given jobId doesn't exist, it initializes the job and then calls the #onData method again.
+ * If the request method is terminate, it terminates the job. Otherwise, it writes the request to the job's stream.
+ *
+ * The #initializeJob method creates a new iframe and sets up a stream to communicate with it.
+ * It also sets up a listener on the stream to write messages from the iframe to the parent and handle job deletion.
+ *
+ * The #handleJobDeletion method checks if a job should be terminated next and if so, it terminates the job.
+ *
+ * The #terminateJob method terminates a job with a given ID.
+ * It removes the iframe associated with the job, destroys the job's stream, and deletes the job from the #jobs record.
+ */
+
+import { WindowPostMessageStream } from '@metamask/post-message-stream';
+import { createWindow, logError } from '@metamask/snaps-utils';
+import type { JsonRpcRequest } from '@metamask/utils';
+import { assert } from '@metamask/utils';
+
+import type { ProxyMessageStream } from './ProxyMessageStream';
+
+type IJob = {
+  id: string;
+  window: Window;
+  stream: WindowPostMessageStream;
+  terminateNext?: string;
+};
+
+type ExecutionControllerArgs = {
+  proxyService: ProxyMessageStream;
+};
+
+/**
+ * The URL of the iframe execution environment.
+ * TODO: This should be configurable, received via params.
+ */
+const IFRAME_URL =
+  'https://metamask.github.io/iframe-execution-environment/0.11.1';
+
+/**
+ * A snap executor using the Webview API.
+ *
+ * This is not a traditional snap executor, as it does not execute snaps itself.
+ * Instead, it creates an iframe window for each snap execution, and sends the
+ * snap execution request to the iframe window. The iframe window is responsible
+ * for executing the snap.
+ *
+ * Webviews can only have a single document, so this executor is
+ * persisted between snap executions. The Webview snap executor essentially
+ * acts as a proxy between the extension and the iframe execution environment.
+ *
+ * @see https://github.com/react-native-webview/react-native-webview
+ */
+export class WebviewSnapExecutor {
```

Comment:
I'm curious what differences there are with the OffscreenSnapExecutor. It seems almost identical except for the stream implementations? 

Perhaps we could just simply have a `ProxySnapExecutor` that both of them use.

Reply:
all comments got resolved on this [commit](https://github.com/MetaMask/snaps/pull/2005/commits/5993c39437a1c66ad58143b60c0854739ea295bd)